### PR TITLE
fix(lsp): normalize windows file URLs properly

### LINKED ISSUE #9744: lsp: Extension reports spurious errors on Windows because of path encoding issues
(EDIT: See comments below, this goes beyond import maps, there's a mismatch in how paths are encoded between the extension and the LSP on Windows)

I'm using an import map to remap `.js` imports to `.ts` in code I share between the browser and Deno.  
It is working fine in Deno, but the extension reports the following errors:

    Unable to load a local module: "file:///c%3A/vscode-deno-import-map/browser/secondary.js".
    Please check the file path.

and

    Property 'value' does not exist on type 'typeof import("file:///c%3A/vscode-deno-import-map/browser/shared")'.

**To Reproduce**

See screenshot below or minimal repro here: https://github.com/elisee/vscode-deno-import-map 

**Expected behavior**

The import map should be properly applied by the extension and there should be no errors.

**Screenshots**

image

**Versions**

vscode: v1.53.2
deno: v1.8.0
extension: v3.1.0
os: windows (seems important, see comment below)

### PR DESCRIPTION
**Pull request by kitsonk**
Fixes #9744
Fixes https://github.com/denoland/vscode_deno/issues/386

### CODE REVIEW

**Comment by kitsonk** [comment]
ok, I have test this under windows and it resolves the issue...

File: cli/lsp/urls.rs [hunk 1]
```diff
@@ -103,12 +103,32 @@ impl LspUrlMap {
...
+      let mut url = url.clone();
+      if url.scheme() == "file" && url.as_str().contains("%3A") {
+        let path = url
+          .path_segments()
+          .unwrap()
+          .map(|s| {
+            if s.ends_with("%3A") && s.len() == 4 {
+              percent_encoding::percent_decode_str(s)
+                .decode_utf8_lossy()
+                .to_string()
```
Comment (piscisaureus) [inline cli/lsp/urls.rs:121]: This `to_string()` is unnecessary.

File: cli/lsp/urls.rs [hunk 2]
```diff
@@ -103,12 +103,32 @@ impl LspUrlMap {
...
       specifier.clone()
     } else {
-      url.clone()
+      let mut url = url.clone();
+      if url.scheme() == "file" && url.as_str().contains("%3A") {
+        let path = url
+          .path_segments()
+          .unwrap()
+          .map(|s| {
+            if s.ends_with("%3A") && s.len() == 4 {
```
Comment (piscisaureus) [inline cli/lsp/urls.rs:118]: * Suggestion: apply only to the first path segment, and only if the url has no (/empty) hostname.
* I don't quite understand why this is necessary. Does "generic" normalization not apply to `%3A`.
* What about `file:///c%3a/foo` (lowecase a) and `file:///c|/foo` (alternative notation) ?

**Comment by piscisaureus** [comment]
It's not quite clear to me what the purpose of normalizing is here. I would suppose that `file:///` URLs would always get percent-decoded, as they are URLs after all. 

So  if I had written: 

`import "file:///C%2A/foo/bar/baz/file%20with%20%25%20and%20%23%20in%20the%20name.txt";`

It'd import from:

`C:\foo\bar\baz\file with % and # in the name.txt`

File: cli/lsp/urls.rs (hunk 1 above)
Comment (kitsonk) [reply cli/lsp/urls.rs:121]: 👍

**Comment by kitsonk** [comment]
@piscisaureus thanks for the feedback.  I have taken a different approach, which should be far more foolproof.  File URLs are transformed into file paths and then back into file URLs.  This ensures that the URL will be normalized in a way that matches Rust's behaviour.  I tested it again on windows and it fixes both the linked issues, but should be able to be generically applied to any client URL.

**Comment by piscisaureus** [comment]
This looks good! Would be great to have a unit test.

**Comment by kitsonk** [comment]
> This looks good! Would be great to have a unit test.

There are two unit tests, one for a Windows system (which ensures the drive letter get recapitalised) and one for non-windows which deals just with percent encoding of parts of the path.

Did you mean some sort of integration test?  Those are complicated in this case because of the need to write out real files, mangle them, craft the messages and then send them in and deal with the nature of windows versus non-windows.

**Review by piscisaureus — APPROVED** [review]
LGTM

**Comment by piscisaureus** [comment]
Sorry, I must have overlooked the tests that were already there.

**Pull request merged**